x/poolmanager/router_test.go

```diff
@@ -1728,6 +1730,484 @@ func (s *KeeperTestSuite) TestSingleSwapExactAmountIn() {
 	}
 }
 
+func (s *KeeperTestSuite) TestEstimateTradeBasedOnPriceImpact() {
+	poolId := uint64(1)
+	maxPriceImpact := sdk.MustNewDecFromStr("0.01")        // 1%
+	maxPriceImpactHalved := sdk.MustNewDecFromStr("0.005") // 0.5%
+	maxPriceImpactTiny := sdk.MustNewDecFromStr("0.0005")  // 0.05%
+
+	externalPriceOneBalancer := sdk.MustNewDecFromStr("0.666666667")   // Spot Price
+	externalPriceTwoBalancer := sdk.MustNewDecFromStr("0.622222222")   // Cheaper than spot price
+	externalPriceThreeBalancer := sdk.MustNewDecFromStr("0.663349917") // Transform adjusted max price impact by 50%
+
+	externalPriceOneStableSwap := sdk.MustNewDecFromStr("1.00000002")             // Spot Price
+	externalPriceTwoStableSwap := sdk.MustNewDecFromStr("0.98989903")             // Cheaper than spot price
+	externalPriceThreeStableSwap := sdk.MustNewDecFromStr("0.990589420505200594") // Transform adjusted max price impact by a %
+
+	externalPriceOneConcentrated := sdk.MustNewDecFromStr("0.0002")                     // Same as spot price 1/5000.000000000000000129
+	externalPriceOneConcentratedInv := sdk.MustNewDecFromStr("5000.000000000000000129") // Inverse of externalPriceOneConcentrated
+	externalPriceTwoConcentrated := sdk.MustNewDecFromStr("0.000198")                   // Cheaper than spot price
+	externalPriceThreeConcentrated := sdk.MustNewDecFromStr("0.000198118")
+
+	assetBaz := "baz"
+	assetBar := "bar"
+	assetUsdc := "usdc"
+	assetEth := "eth"
+
+	clCoinsLiquid := sdk.NewCoins(
+		sdk.NewCoin("eth", sdk.NewInt(1000000)),
+		sdk.NewCoin("usdc", sdk.NewInt(5000000000)),
+	)
+	clCoinsNotLiquid := sdk.NewCoins(
+		sdk.NewCoin("eth", sdk.NewInt(1)),
+		sdk.NewCoin("usdc", sdk.NewInt(1)),
+	)
+
+	// The below values have been tested and hard coded by using the `CalcOutAmtGivenIn` as it is quite hard to
+	// mathematically work these out.
+	tests := map[string]struct {
```

Comment:
Can we add test cases for 
- Stable swap panic error catching
- adjustedMaxPriceImpact is zero or negative
- tokenOut from calcOutAmtGivenIn is zero
-